services/src/main/java/org/keycloak/protocol/saml/DefaultSamlArtifactResolver.java

```diff
@@ -0,0 +1,187 @@
+        String realmId = userSession.getRealm().getId();
+        return buildArtifact(realmId, entityId,samlDocument);
+    }
+
+    @Override
+    public String buildAuthnArtifact(String entityId, Document samlDocument, AuthenticatedClientSessionModel clientSession) throws ArtifactResolverProcessingException, ArtifactResolverConfigException{
+        String realmId = clientSession.getRealm().getId();
+        return buildArtifact(realmId, entityId,samlDocument);
+    }
+
+    private String buildArtifact(String realmId, String entityId, Document samlDocument) throws ArtifactResolverProcessingException, ArtifactResolverConfigException {
+        try {
+            String artifact = createArtifact(entityId);
+            String artifactResponseStr = DocumentUtil.getDocumentAsString(samlDocument);
+            session.sessions().addArtifactResponse(realmId, artifact, artifactResponseStr);
+            return artifact;
+        }catch(ProcessingException e){
+            throw new ArtifactResolverProcessingException(e);
+        }catch(ConfigurationException e) {
+            throw new ArtifactResolverConfigException(e);
+        }
+    }
+
+    @Override
+    public String resolveArtifact(String artifact) throws ArtifactResolverProcessingException {
+        String res = session.sessions().getArtifactResponse(artifact);
+
+        if (res == null ){
+            throw new ArtifactResolverProcessingException("Cannot find artifact "+ artifact + " in cache");
+        }
+        session.sessions().removeArtifactResponse(artifact);
+        return res;
+    }
+
+    @Override
+    public void initialize(KeycloakSession session) {
+        this.session = session;
+    }
+
+
+    @Override
+    public ClientModel selectSourceClient(String artifact, List<ClientModel> clients) throws ArtifactResolverProcessingException {
+        try {
+            byte[] source = extractSourceFromArtifact(artifact);
+
+            MessageDigest sha1Digester = MessageDigest.getInstance("SHA-1");
+
+            for (ClientModel client :clients) {
+                byte[] clientBytes = sha1Digester.digest(client.getClientId().getBytes(Charsets.UTF_8));
+                if (Arrays.equals(source, clientBytes)) {
+                    return client;
+                }
+            }
+        } catch(NoSuchAlgorithmException e) {
+            throw new ArtifactResolverProcessingException(e);
+        }
+
+        throw new ArtifactResolverProcessingException("No client matching the artifact source found");
+    }
+
+    private void assertSupportedArtifactFormat(String artifactString) throws ArtifactResolverProcessingException{
+        byte[] artifact = Base64.getDecoder().decode(artifactString);
+
+        if (artifact.length != 44) {
+            throw new ArtifactResolverProcessingException("Artifact " + artifactString + " has a length of " + artifact.length  + ". It should be 44");
+        }
+        if (artifact[0] != 0 || artifact[1] != 4) {
```

Comment:
That's why there is a SPI. If someone want something else he can code it, but the default version (04) follows a standard described in SAML. 
It should be seen as a whole block, if we change the version, it would change the way of computing the artifact so I don't see any benefit of creating an Abstract class here. Am I missing something ?